### PR TITLE
Type safety in composition time

### PR DESCRIPTION
**Pull request by uriva**

### CODE REVIEW

**Comment by uriva** [comment]
automerge

File: computation_graph/type_safety_test.py [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+        pass
+
+    def g(x: str) -> str:
+        pass
+
+    assert type_safety.can_compose(f, g, "y")
+    assert not type_safety.can_compose(f, g, "x")
+
+
+def test_is_subtype():
```
Comment (nitzo) [inline computation_graph/type_safety_test.py:38]: consider parameterizing

File: computation_graph/type_safety_test.py [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+        [Set[str], Collection[str]],
+        [List, Sequence],
+        [Union[int, str], Union[int, str]],
+        [str, Union[int, str]],
+        [Union[List, Set], Collection],
+    ]:
+        assert type_safety.is_subtype(x)
+
+
+def test_not_is_subtype():
```
Comment (nitzo) [inline computation_graph/type_safety_test.py:52]: ditto

**Review by nitzo — APPROVED** [review]

**Comment by uriva** [comment]
add support for `Optional`

**Pull request merged**